tests/AgDatabaseMove.Unit/BackupOrder.cs

```diff
@@ -44,51 +60,161 @@ private static List<BackupMetadata> ListBackups()
+      var listWithStripes = CloneBackupMetaDataList(list).ToList();
+      listWithStripes.ForEach(b => {
+        var path = b.PhysicalDeviceName.Split('.');
+        b.PhysicalDeviceName = $"{path[0]}_striped.{path[1]}";
+      });
+      list.AddRange(listWithStripes);
+      return list;
+    }
+
+    private static List<BackupMetadata> GetBackupListWithStripesAndDuplicates()
+    {
+      var listWithStripes = GetBackupListWithStripes();
+      var duplicate = CloneBackupMetaDataList(listWithStripes);
+      listWithStripes.AddRange(duplicate);
+      return listWithStripes;
+    }
+
+    private static List<BackupMetadata> GetBackupListWithoutFull()
+    {
+      var list = GetBackupList();
+      list.RemoveAll(b => b.BackupType == BackupFileTools.BackupType.Full);
+      return list;
+    }
+
+    private static void VerifyListIsAValidBackupChain(List<BackupMetadata> backupChain)
+    {
+      var listOfFileNames = backupChain.Select(b => b.PhysicalDeviceName).ToList();
+      Assert.True(listOfFileNames.Count == listOfFileNames.Distinct().Count(), "There are no duplicates?");
+
+      var fullBackup = backupChain.First();
+      Assert.NotNull(fullBackup);
+      Assert.True(fullBackup.BackupType == BackupFileTools.BackupType.Full, "Is the first backup a full backup?");
+      backupChain.RemoveAt(0);
+
+      BackupMetadata currentBackup;
+      var prevBackup = fullBackup;
+
+      while((currentBackup = backupChain.FirstOrDefault())?.BackupType == BackupFileTools.BackupType.Full) {
+        Assert.True(currentBackup.CheckpointLsn == prevBackup.CheckpointLsn,
+                    "Is it a striped file of the full backup?");
+        prevBackup = currentBackup;
+        backupChain.RemoveAt(0);
+      }
+
+      prevBackup = currentBackup;
+      while((currentBackup = backupChain.FirstOrDefault())?.BackupType == BackupFileTools.BackupType.Diff) {
+        Assert.True(currentBackup.DatabaseBackupLsn == fullBackup.CheckpointLsn,
+                    "Is the diff linked to the full backup?");
+        Assert.True(currentBackup.FirstLsn == prevBackup.FirstLsn && currentBackup.LastLsn == prevBackup.LastLsn,
+                    "Is it a striped file of the diff backup?");
+        prevBackup = currentBackup;
+        backupChain.RemoveAt(0);
+      }
+
+      prevBackup = currentBackup;
+      while((currentBackup = backupChain.FirstOrDefault())?.BackupType == BackupFileTools.BackupType.Log) {
+        Assert.True(currentBackup.DatabaseBackupLsn == fullBackup.CheckpointLsn,
+                    "Is the log linked to the full backup?");
+        Assert.True(
+                    (currentBackup.FirstLsn == prevBackup.FirstLsn && currentBackup.LastLsn == prevBackup.LastLsn) ||
+                    (prevBackup.LastLsn == currentBackup.FirstLsn),
+                    "Is it either the striped file of the prev log backup or the next one in the chain?");
+
+        prevBackup = currentBackup;
+        backupChain.RemoveAt(0);

```

Comment:
This test has 3 while loops... I think you can simplify it down into 1 with some booleans to help:
```C#
bool foundFull, foundDiff, foundLog = false;
while(curr = backupChain.FirstOrDefault())
{
  if(curr.BackupType == Full)
  {
    if (foundDiff || foundLog)
       Assert.False():
    foundFull = true;
    backupChain.Remove(b => b.CheckPointLSN == curr.CheckpointLSN)
  }
  if(curr.BackupType == Diff)
   {
       if (foundDiff || foundLog || !foundFull)
         Assert.False():
      backupChain.Remove(b => b.LastLSN== curr.LastLSN)
   }
  if(curr.BackupType == Log)
  {
    if (!foundFull)
         Assert.False():
    backupChain.Remove(b => b.LastLSN== curr.LastLSN)
  }
}


```

Reply:
Ah yes, thank you - I think I was stuck in a pattern of thinking we need to handle each in sequence
Rather we can handle it more like "when we get this type -> Assert this, this and that" 

This almost works - just need to add in the ordering/link checks for the logs and diff (i.e. they all point to the `full.checkpointLsn` and that for each `prev.lastLsn==curr.firstLns`)
Will implement it!